core/modules/map.rs

```diff
@@ -1301,6 +1302,58 @@ impl ModuleMap {
 
     vec![]
   }
+
+  /// Load and evaluate an ES module provided the specifier and source code.
+  ///
+  /// The module should not have Top-Level Await (that is, it should be
+  /// possible to evaluate it synchronously).
+  ///
+  /// It is caller's responsibility to ensure that not duplicate specifiers are
+  /// passed to this method.
+  pub(crate) fn lazy_load_es_module_from_code(
+    &self,
+    scope: &mut v8::HandleScope,
+    module_specifier: &str,
+    source_code: ModuleCode,
+  ) -> Result<v8::Global<v8::Value>, Error> {
+    let specifier = ModuleSpecifier::parse(module_specifier)?;
+    let mod_id = self
+      .new_es_module(scope, false, specifier.into(), source_code, false)
+      .map_err(|e| match e {
+        ModuleError::Exception(exception) => {
+          let exception = v8::Local::new(scope, exception);
+          exception_to_err_result::<()>(scope, exception, false).unwrap_err()
+        }
+        ModuleError::Other(error) => error,
+      })?;
+
+    self.instantiate_module(scope, mod_id).map_err(|e| {
+      let exception = v8::Local::new(scope, e);
+      exception_to_err_result::<()>(scope, exception, false).unwrap_err()
+    })?;
+
+    let module_handle = self.get_handle(mod_id).unwrap();
+    let module_local = v8::Local::<v8::Module>::new(scope, module_handle);
+
+    let status = module_local.get_status();
+    assert_eq!(status, v8::ModuleStatus::Instantiated);
+
+    let value = module_local.evaluate(scope).unwrap();
+    let promise = v8::Local::<v8::Promise>::try_from(value).unwrap();
+    let result = promise.result(scope);
+    if !result.is_undefined() {
+      return Err(
+        exception_to_err_result::<()>(scope, result, false).unwrap_err(),
+      );
+    }
+
+    let status = module_local.get_status();
+    assert_eq!(status, v8::ModuleStatus::Evaluated);
```

Comment:
Maybe return an `Err` if module has to be asyncronously evaluated (i.e contains top level async)